talisker/logs.py

```diff
@@ -359,36 +365,63 @@ def logfmt_key(self, k):
         if isinstance(k, bytes):
             k = k.decode('utf8')
 
-        # TODO: look at optimizing this to avoid string allocations
+        if isinstance(k, str):
+            k = k.replace(' ', '_').replace('.', '_').replace('=', '_')
+            k = self.safe_string(k, self.MAX_KEY_SIZE, self.TRUNCATED)
+            # TODO: look at measuring perf of this
+            # ' ' and = are replaced because they're are not valid logfmt
+            # . is replaced because elasticsearch can't do keys with . in
+        elif isinstance(k, numbers.Number):
+            k = str(k)
+        else:
+            return None
 
-        k = k.strip()[:self.MAX_KEY_SIZE]
-        # ' ' and = are replaced because they're are not valid logfmt
-        # . is replaced because elasticsearch can't do keys with . in
-        k = k.replace(' ', '_').replace('.', '_').replace('=', '_')
-        # strip " as grok parser can not escape them
-        k = k.replace('"', '')
         return k
 
     def logfmt_value(self, v):
-        if isinstance(v, bytes):
-            v = v.decode('utf8')
+        """Format a python value for logfmt.
 
-        if isinstance(v, str):
-            v = v.strip()
-            # logstash kv filter cannot handle escaped "
-            v = v.replace('"', '')
+        The output target here is influenced by logstash/elastic search types.
 
-            if len(v) > self.MAX_VALUE_SIZE:
-                v = v[:self.MAX_VALUE_SIZE] + self.TRUNCATED
+        For strings, ensures unicode, quotes, and truncates as needed.
+        For bools, format as json truth values.
+        For dicts, truncate.
+        For others types (e.g. numbers) leave as is.
+        For all else, coerce to string.
+        """
 
+        if isinstance(v, bytes):
+            v = v.decode('utf8')
+        if isinstance(v, str):
+            v = self.safe_string(v, self.MAX_VALUE_SIZE, self.TRUNCATED)
+            # explicitly quote strings to avoid ambiguity
             v = '"' + v + '"'
         elif isinstance(v, bool):
             v = str(v).lower()
-        elif isinstance(v, dict):
-            return '"' + self.TRUNCATED + '"'
+        elif isinstance(v, numbers.Number):
+            v = str(v)
+        else:
+            v = '"' + str(type(v)) + '"'
 
         return v
 
+    def safe_string(self, s, max, truncate):
+        s = s.strip()
+
+        # no new lines allowed, so truncate at the first new line
+        if '\n' in s:
+            s = s.split('\n', 1)[0] + truncate
```

Comment:
hmm, if the first line is longer than the maximum length then we'll add truncate twice!